proxy-socket-core/src/main/java/net/airvantage/proxysocket/core/v2/ProxyProtocolV2Decoder.java

```
@@ -0,0 +1,192 @@
+            throws ProxyProtocolParseException {
+        if (variableLength < 2*(IPV6_ADDR_LEN + PORT_LEN)) {
+            throw new ProxyProtocolParseException("Truncated IPv6 address block in header");
+        }
+
+        InetAddress s;
+        InetAddress d;
+        byte[] sb = new byte[16];
+        byte[] db = new byte[16];
+        System.arraycopy(data, pos, sb, 0, IPV6_ADDR_LEN);
+        System.arraycopy(data, pos+IPV6_ADDR_LEN, db, 0, IPV6_ADDR_LEN);
+        try {
+            s = InetAddress.getByAddress(sb);
+            d = InetAddress.getByAddress(db);
+        } catch (UnknownHostException e) {
+            throw new ProxyProtocolParseException("Invalid IPv6 address in header", e);
+        }
+
+        pos += 2*IPV6_ADDR_LEN;
+        int sp = ((data[pos++] & 0xFF) << 8) | (data[pos++] & 0xFF);
+        int dp = ((data[pos++] & 0xFF) << 8) | (data[pos++] & 0xFF);
+
+        return new AddressPair(new InetSocketAddress(s, sp), new InetSocketAddress(d, dp), pos);
+    }
+
+    private static AddressPair parseUnixAddresses(byte[] data, int pos, int variableLength)
+            throws ProxyProtocolParseException {
+        if (variableLength < 2*UNIX_ADDR_LEN) {
+            throw new ProxyProtocolParseException("Truncated UNIX address block in header");
+        }
+        pos += 2*UNIX_ADDR_LEN;
+        throw new ProxyProtocolParseException("UNIX Address Processing not implemented");
+    }
+
+    private static List<Tlv> parseTlvs(byte[] data, int pos, int tlvLen) {
+        List<Tlv> tlvs = new ArrayList<>();
+        int tlvPos = pos;
+        int tlvEnd = tlvPos + tlvLen;
+        while (tlvPos + TLV_HEADER_LEN <= tlvEnd) {
+            int type = data[tlvPos++] & 0xFF;
+            int len = ((data[tlvPos++] & 0xFF) << 8) | (data[tlvPos++] & 0xFF);
+
+            if (tlvPos + len > tlvEnd) break;
+            tlvs.add(new Tlv(type, data, tlvPos, len));
+            tlvPos += len;
+        }
```

if there is remaining byte we should raise an exception ?